File: sp/src/main/java/wut/sp/Interfaces/IElement.java
package wut.sp.Interfaces;

import java.util.List;
import java.util.ArrayList;

public abstract class IElement {
    protected List<IElement> elementList;
    
    public IElement() {
        this.elementList = new ArrayList<IElement>();
    }

    public void add(IElement element) {
        this.elementList.add(element);
    }

    public void remove(IElement element) {
        this.elementList.remove(element);
    }

    public IElement get(int index) {
        return this.elementList.get(index);
    }

    public abstract void print();
    public abstract IElement clone();
}

File: spC++/src/Implement/IElement.cpp
#include "../Headers/IElement.h"

void IElement::add(IElement* element) {
    children.push_back(element);
}

void IElement::remove(IElement* element) {
    children.remove(element);
}

std::list<IElement*> IElement::get(int index) {
    std::list<IElement*> result;
    for (IElement* element : children) {
        result.push_back(element);
    }
    return result;
}

void *IElement::accept(IElement *visitee) {
    return visitee->accept(this);
}